Automattic/wp-calypso
# CI: Overhaul for speed 🚀

This is a structural overhaul to the CircleCI setup. The goal was speed, leveraging CI parallelism and caching to make our CI runs as fast as possible 🚀 

Evidence is anecdotal, metrics from CI aren't great as far as I can tell, but `master` seems to run in 2-3 minutes. This branch's workflow runs 3 parallel jobs. `test-client`, the slowest job, tends to run in around a minute.

- Use shared references. This will helps to share common setup among jobs.
- Add eslint report to test summary. Because of eslint-eslines, we lose the inline report in exchange for the CI-based report.
- Inline CI test runners. The indirection through package.json or bins make it more difficult to work with the CI config. The indirection complicates things like using environment variables in the commands.
- Remove unneeded config and scripts now that CI test scripts are inlined.
- Cache the Jest cache.
- I explored caching the Babel cache, but that didn't seem to improve build time (e988d96fdb7b2ed8123b7d55c3c103ada2b2a4f1)
- I explored [`--changedSince=master`](https://jestjs.io/docs/en/cli#changedsince), but any difference was negligible. It's likely that the time taken to build out the dependency graph for `findRelatedTests` on every container negates any improvement.
- Add [`--ci`](https://jestjs.io/docs/en/cli#ci) Jest option
  > When this option is provided, Jest will assume it is running in a CI environment. This changes the behavior when a new snapshot is encountered. Instead of the regular behavior of storing a new snapshot automatically, it will fail the test and require Jest to be run with `--updateSnapshot`.
- Add `--runTestsByPath` Jest option for CircleCI split test. CircleCI splits a glob and passes full paths to Jest. Without this option, Jest will attempt to fuzzy match every full path it is passed.
  > `--runTestsByPath`
  > Used when provided patterns are exact file paths. This avoids converting them into a regular expression and matching it against every single file.
  > [boolean] [default: false]
- Add [`--silent `](https://jestjs.io/docs/en/cli#silent) Jest option
  > Prevent tests from printing messages through the console.
- Put test results directly into results folder for storage. Eliminates the need for a "gather results" step.
- Split sequential steps into a parallel workflow with different jobs:
  - `lint-and-translate`: Run eslint, generate translations. These jobs have been kept sequential because they appear to be fast enough that the overhead to startup 2 jobs is slower.
  - `test-client`: Run the client tests. This is the only job that seems worthwhile to run on parallel containers.
  - `test-server`: Run the server tests.
  - `test-integration`: This job is part of a cron workflow along with `test-client` and `test-server`. It is not run on demand (PR), but at scheduled intervals.
- Add `node_modules` cache. Previously we had eliminated the `node_modules` cache because we were using `npm ci` which deletes the folder (#25431). Then, we maintained this behavior, but began to cache the npm cache so that `npm ci` would execute faster because the modules were already available locally (#25487). This approach restores the `node_modules` cache and _completely skips_ `npm ci` if there was a cache hit. This should be safe because the `node_modules` cache requires a checksum match on `.nvmrc` and `npm-shrinkwrap.json`. It's important to ensure that _all variables_ which might generate different `node_modules` structure are part of the cache key, e.g. running on different architectures.
- Update the `git` cache with branch and revision. With a static key, the cache lives forever and is never updated.
- ~Add [`cache-require-paths`](https://www.npmjs.com/package/cache-require-paths) to the client and server Jest tests. In some anecdotal testing this seemed to provide improvement, but we might roll this back for now and do some more isolated testing after landing the bulk of this PR.~
- Quote shell variables

## Failing tests example

Here's an example of forced expected failure:

https://circleci.com/workflow-run/1ccbd0c0-19e2-4cc0-8de4-6ddfdf120d3a

![summary](https://user-images.githubusercontent.com/841763/45611698-98eb8a00-ba60-11e8-910b-f03f0ceb3c00.png)

Note that although lint fails, the sequential translate tasks still run.

This is the lint error summary that did not exist previously:

![lint-fail](https://user-images.githubusercontent.com/841763/45611668-76f20780-ba60-11e8-87c1-a3155fc76a3b.png)


## Testing
- Is CI working as expected?
- Do local test runs continue to work as expected?
  - `npm run test-client`
  - `npm run test-server`
  - `npm run test-integration`
- Is CI running faster? 😁 

## Follow-up
- Remove non CI-based cron to run integration tests
- `projects` based Jest running when monorepo begins (#27123)
- Pull e2e tests into this repo? This lays the groundwork…

From #27150
